### PR TITLE
Sparse constraint matrix analysis

### PR DESCRIPTION
**Pull request by guykatzz**
The dense matrix approach was becoming a bottleneck for larger networks

### CODE REVIEW

File: src/engine/Engine.cpp [hunk 1]
```diff
@@ -773,7 +773,7 @@ void Engine::removeRedundantEquations( const double *constraintMatrix )
     analyzer->analyze( constraintMatrix, m, n );
 
     ENGINE_LOG( Stringf( "Number of redundant rows: %u out of %u",
-                  analyzer->getRedundantRows().size(), m ).ascii() );
+                         analyzer->getRedundantRows().size(), m ).ascii() );
```
Comment (anwu1219) [inline src/engine/Engine.cpp:776]: remove spaces

File: src/engine/Engine.cpp [hunk 2]
```diff
@@ -2105,11 +2105,3 @@ void Engine::setConstraintViolationThreshold( unsigned threshold )
 {
     _smtCore.setConstraintViolationThreshold( threshold );
 }
-
```
Comment (anwu1219) [inline src/engine/Engine.cpp:2108]: Why are these lines removed?

Comment (guykatzz) [reply src/engine/Engine.cpp:2108]: Leftovers from when we were using Makefiles (and not CMake)

File: src/engine/Engine.cpp (hunk 1 above)
Comment (guykatzz) [reply src/engine/Engine.cpp:776]: The new spaces? It was not properly indented before, now it is directly below the "

File: src/engine/ConstraintMatrixAnalyzer.h [hunk 3]
```diff
@@ -27,48 +28,57 @@ class ConstraintMatrixAnalyzer : public IConstraintMatrixAnalyzer
...
+
+    unsigned _pivotRow;
+    unsigned _pivotColumn;
+    double _pivotElement;
+
+    double *_workRow;
+    double *_workRow2;
+
     /*
       The i'th (permuted) column of the matrix is stored in memory
```
Comment (anwu1219) [inline src/engine/ConstraintMatrixAnalyzer.h:61]: ith (permuted) row?
